go/test/endtoend/bootstrap_test.go

```diff
@@ -0,0 +1,549 @@
+	require.NoError(t, os.MkdirAll(etcdDataDir, 0o755))
+
+	etcdClientAddr, etcdCmd := startEtcd(t, etcdDataDir)
+	defer func() {
+		if etcdCmd != nil && etcdCmd.Process != nil {
+			_ = etcdCmd.Process.Kill()
+			_ = etcdCmd.Wait()
+		}
+	}()
+
+	t.Logf("Started etcd at %s", etcdClientAddr)
+
+	// Create topology server and cell
+	testRoot := "/multigres"
+	globalRoot := filepath.Join(testRoot, "global")
+	cellName := "test-cell"
+	cellRoot := filepath.Join(testRoot, cellName)
+
+	ts, err := topo.OpenServer("etcd2", globalRoot, []string{etcdClientAddr})
+	require.NoError(t, err, "Failed to open topology server")
+	defer ts.Close()
+
+	err = ts.CreateCell(ctx, cellName, &clustermetadatapb.Cell{
+		ServerAddresses: []string{etcdClientAddr},
+		Root:            cellRoot,
+	})
+	require.NoError(t, err, "Failed to create cell")
+
+	// Use postgres database (multigres always uses postgres database with table_group for isolation)
+	database := "postgres"
+	err = ts.CreateDatabase(ctx, database, &clustermetadatapb.Database{
+		Name:             database,
+		DurabilityPolicy: "ANY_2",
+	})
+	require.NoError(t, err, "Failed to create database in topology")
+
+	t.Logf("Created database '%s' with policy 'ANY_2'", database)
+
+	// Create 3 empty nodes
+	shardID := "test-shard-01"
+	nodes := make([]*nodeInstance, 3)
+	for i := 0; i < 3; i++ {
+		nodes[i] = createEmptyNode(t, tempDir, cellName, shardID, database, i, etcdClientAddr)
+		defer cleanupNode(t, nodes[i])
+	}
+
+	t.Logf("Created 3 empty nodes")
+
+	// Verify all nodes are uninitialized
+	for i, node := range nodes {
+		status := checkInitializationStatus(t, node)
+		require.False(t, status.IsInitialized, "Node %d should be uninitialized", i)
+		t.Logf("Node %d (%s) confirmed uninitialized", i, node.name)
+	}
+
+	// Create coordinator nodes for bootstrap action
+	rpcClient := rpcclient.NewClient(10) // connection pool capacity
+	coordNodes := make([]*coordinator.Node, 3)
+	for i, node := range nodes {
+		pooler := &clustermetadatapb.MultiPooler{
+			Id: &clustermetadatapb.ID{
+				Component: clustermetadatapb.ID_MULTIPOOLER,
+				Cell:      node.cell,
+				Name:      node.name,
+			},
+			Hostname: "localhost",
+			PortMap: map[string]int32{
+				"grpc": int32(node.grpcPort),
+			},
+			Shard:    shardID,
+			Database: database,
+		}
+
+		coordNodes[i] = &coordinator.Node{
+			ID:        pooler.Id,
+			Hostname:  pooler.Hostname,
+			Port:      int32(node.pgPort),
+			ShardID:   shardID,
+			RpcClient: rpcClient,
+			Pooler:    pooler,
+		}
+	}
+
+	// Execute bootstrap action
+	t.Logf("Executing bootstrap action...")
+	logger := slog.Default()
+	bootstrapAction := actions.NewBootstrapShardAction(ts, logger)
+
+	err = bootstrapAction.Execute(ctx, shardID, database, coordNodes)
+	require.NoError(t, err, "Bootstrap action should succeed")
+
+	t.Logf("Bootstrap action completed successfully")
+
+	// Wait for initialization to complete
+	time.Sleep(2 * time.Second)
```

Comment:
FWIW, my personal preference is to favor require.Eventually() over time.Sleep() where possible so there's a possibility it doesn't need to take the entire 2 seconds and tests could potentially run faster